src/box/lua/misc.cc

```diff
@@ -197,6 +197,68 @@ lua_wrap_formatted_array(struct lua_State *L, struct tuple_format *format)
 	lua_setmetatable(L, -2);
 }
 
+/**
+ * Advances the iterator - invokes iterator_next() with a saved state.
+ * For details, see description of port_c_push_iterator_lua().
+ */
+static int
+port_c_iterator_lua_next(lua_State *L)
+{
+	struct port_c_iterator *iter =
+		(struct port_c_iterator *)lua_touserdata(L, 1);
+	bool is_eof = false;
+	struct port port;
+	int rc = iter->next(iter, &port, &is_eof);
+	if (rc != 0)
+		luaT_error(L);
+	if (is_eof)
+		return 0;
+	int top_svp = lua_gettop(L);
+	port_dump_lua(&port, L, PORT_DUMP_LUA_MODE_FLAT);
+	port_destroy(&port);
+	return lua_gettop(L) - top_svp;
+}
+
+/**
+ * The function should be called as a closure with one upvalue:
+ * port_c_iterator stored as userdata.
+ * For details, see description of port_c_push_iterator_lua().
+ */
+static int
+port_c_iterator_lua_start(lua_State *L)
+{
+	lua_pushvalue(L, lua_upvalueindex(1));
+	return 1;
+}
+
+/**
+ * Pushes an iterator created by passed iterable object to Lua stack.
+ *
+ * Iterators in Lua are implemented as usual functions (or closures), which
+ * return the next element. So this function pushes a closure, that returns an
+ * actual iterator - a callable userdata, which is a wrapper over iterator_next.
+ * That's how it looks from Lua:
+ * function(iter)
+ *     for v1, v2 in iter() do
+ *         process(v1, v2)
+ *     end
+ * end
+ */
+static void
+port_c_push_iterator_lua(struct lua_State *L, port_c_iterable *iterable)
+{
+	struct port_c_iterator *iter =
+		(struct port_c_iterator *)lua_newuserdata(L, sizeof(*iter));
+	lua_createtable(L, 0, 1);
+	lua_pushcfunction(L, port_c_iterator_lua_next);
+	lua_setfield(L, -2, "__call");
```

Comment:
It's probably better to create one metatable and register a new userdata type for all iterator objects. It should be more efficient this way.

Reply:
Done.